servicetalk-opentracing-zipkin-publisher/src/main/java/io/servicetalk/opentracing/zipkin/publisher/reporter/HttpReporter.java

```diff
@@ -0,0 +1,252 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.opentracing.zipkin.publisher.reporter;
+
+import io.servicetalk.buffer.api.Buffer;
+import io.servicetalk.buffer.api.BufferAllocator;
+import io.servicetalk.concurrent.CompletableSource;
+import io.servicetalk.concurrent.PublisherSource;
+import io.servicetalk.concurrent.api.AsyncCloseable;
+import io.servicetalk.concurrent.api.BufferStrategy.Accumulator;
+import io.servicetalk.concurrent.api.Completable;
+import io.servicetalk.concurrent.api.CompositeCloseable;
+import io.servicetalk.concurrent.api.Publisher;
+import io.servicetalk.http.api.HttpClient;
+import io.servicetalk.http.api.SingleAddressHttpClientBuilder;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import zipkin2.Component;
+import zipkin2.Span;
+import zipkin2.codec.SpanBytesEncoder;
+import zipkin2.reporter.Reporter;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import javax.annotation.Nonnull;
+
+import static io.servicetalk.concurrent.api.AsyncCloseables.newCompositeCloseable;
+import static io.servicetalk.concurrent.api.AsyncCloseables.toAsyncCloseable;
+import static io.servicetalk.concurrent.api.BufferStrategies.forCountOrTime;
+import static io.servicetalk.concurrent.api.Completable.completed;
+import static io.servicetalk.concurrent.api.Processors.newCompletableProcessor;
+import static io.servicetalk.concurrent.api.Processors.newPublisherProcessorDropHeadOnOverflow;
+import static io.servicetalk.concurrent.api.SourceAdapters.fromSource;
+import static io.servicetalk.concurrent.api.SourceAdapters.toSource;
+import static io.servicetalk.concurrent.internal.FutureUtils.awaitTermination;
+import static io.servicetalk.http.api.CharSequences.newAsciiString;
+import static io.servicetalk.http.api.HttpHeaderNames.CONTENT_TYPE;
+import static io.servicetalk.http.api.HttpHeaderValues.APPLICATION_JSON;
+import static java.time.Duration.ofSeconds;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A {@link Span} {@link Reporter} that will publish to an HTTP endpoint with a configurable encoding {@link Codec}.
+ */
+public final class HttpReporter extends Component implements Reporter<Span>, AsyncCloseable {
+    private static final Logger LOGGER = LoggerFactory.getLogger(HttpReporter.class);
+    static final String V1_PATH = "/api/v1/spans";
+    static final String V2_PATH = "/api/v2/spans";
+    static final CharSequence THRIFT_CONTENT_TYPE = newAsciiString("application/x-thrift");
+    static final CharSequence PROTO_CONTENT_TYPE = newAsciiString("application/protobuf");
+
+    private final PublisherSource.Processor<Span, Span> buffer;
+    private final AsyncCloseable closeable;
+
+    private volatile boolean closeInitiated;
+
+    private HttpReporter(final Builder builder) {
+        final HttpClient client = builder.clientBuilder.build();
```

Comment:
Consider creating `CompositeCloseable` upfront and appending the client asap. Then try-catch the whole ctor and close the closeable in case of any errors.